syntax = "proto2";

service CalculateService {
  rpc CalculateSomething (stream CalculateSomethingRequest) returns (stream CalculateSomethingResponse);
}

message CalculateSomethingRequest {
  repeated int32 numbers = 1;
  optional OperationType operation_type = 2 [default = MUL];
  required int32 divider = 3;
}

message CalculateSomethingResponse {
  message CalculateSomethingResponseValid {
    required int32 result = 1;
    required OperationType operation_type = 2;
  }

  message CalculateSomethingResponseInvalid {
    required int32 code = 1;
    optional string message = 2;
  }

  oneof result {
    CalculateSomethingResponseValid valid_result = 1;
    CalculateSomethingResponseInvalid invalid_result = 2;
  }

}

enum OperationType {
  ADD = 0;
  MUL = 1;
}